Python: Support serverless environments

### Feature Request / Improvement

## Problem

PyIceberg currently relies on [`multiprocessing`](https://docs.python.org/3/library/multiprocessing.html) in [`Table.plan_files`](https://github.com/apache/iceberg/blob/a73f10b3e7a98d7efcab7e01382b78bffdc7028e/python/pyiceberg/table/__init__.py#L776) and [the pyarrow interface](https://github.com/apache/iceberg/blob/a73f10b3e7a98d7efcab7e01382b78bffdc7028e/python/pyiceberg/io/pyarrow.py#L873C49-L873C49). Unfortunately, `multiprocessing` relies on `/dev/shm`, which is not provided by serverless runtimes like [AWS Lambda](https://aws.amazon.com/blogs/compute/parallel-processing-in-python-with-aws-lambda/) and [AWS Fargate](https://github.com/aws/containers-roadmap/issues/710). In effect, reliance on `/dev/shm` assumes the user has control over the host environment and thus disqualifies use in serverless environments.

[Apparently](https://github.com/lambci/docker-lambda/issues/75#issuecomment-668897239), one way to emulate Lambda or Fargate container runtimes locally is by running a container with [`--ipc="none"`](https://docs.docker.com/engine/reference/run/#ipc-settings---ipc). This will disable the `/dev/shm` mount and cause the `multiprocessing` module to fail with the following error:

```python
[ERROR] OSError: [Errno 38] Function not implemented
Traceback (most recent call last):
  File "/var/task/app.py", line 27, in handler
    result = scan.to_arrow().slice_length(limit)
  File "/var/task/pyiceberg/table/__init__.py", line 819, in to_arrow
    self.plan_files(),
  File "/var/task/pyiceberg/table/__init__.py", line 776, in plan_files
    with ThreadPool() as pool:
  File "/var/lang/lib/python3.10/multiprocessing/pool.py", line 930, in __init__
    Pool.__init__(self, processes, initializer, initargs)
  File "/var/lang/lib/python3.10/multiprocessing/pool.py", line 196, in __init__
    self._change_notifier = self._ctx.SimpleQueue()
  File "/var/lang/lib/python3.10/multiprocessing/context.py", line 113, in SimpleQueue
    return SimpleQueue(ctx=self.get_context())
  File "/var/lang/lib/python3.10/multiprocessing/queues.py", line 341, in __init__
    self._rlock = ctx.Lock()
  File "/var/lang/lib/python3.10/multiprocessing/context.py", line 68, in Lock
    return Lock(ctx=self.get_context())
  File "/var/lang/lib/python3.10/multiprocessing/synchronize.py", line 162, in __init__
    SemLock.__init__(self, SEMAPHORE, 1, 1, ctx=ctx)
  File "/var/lang/lib/python3.10/multiprocessing/synchronize.py", line 57, in __init__
    sl = self._semlock = _multiprocessing.SemLock(
```

Interesting note from [`multiprocessing.pool.ThreadPool`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.ThreadPool):
> A `ThreadPool` shares the same interface as `Pool`, which is designed around a pool of processes and predates the introduction of the `concurrent.futures` module [...] Users should generally prefer to use `concurrent.futures.ThreadPoolExecutor`, which has a simpler interface that was designed around threads from the start, and which returns `concurrent.futures.Future` instances that are compatible with many other libraries, including `asyncio`.

## Proposal

Perhaps PyIceberg should support multiple concurrency strategies, allowing the user to configure which is most appropriate for their runtime/resources.

Instead of using `multiprocessing` directly, we could instead depend on a concrete implementation of an [`Executor`](https://docs.python.org/3/library/concurrent.futures.html#executor-objects) from the [concurrent.futures](https://docs.python.org/3/library/concurrent.futures.html) module wherever we need concurrency. The user can select the appropriate implementation via configuration:

- `PYICEBERG__CONCURRENCY=process` uses [`ProcessPoolExecutor`](https://docs.python.org/3/library/concurrent.futures.html#processpoolexecutor) (default, same as current implementation)
- `PYICEBERG__CONCURRENCY=thread` uses [`ThreadPoolExecutor`](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor) (appropriate for serverless environments)

This might even allow PyIceberg to support other concurrency models in the future, e.g., user-defined implementations of `Executor`.

I reproduced this problem and have a fix on a fork, confirming that this approach at least works. Depending on feedback from the community, I can tidy things up and submit a PR. 😃 

### Query engine

None

--- a/python/pyiceberg/io/pyarrow.py
+++ b/python/pyiceberg/io/pyarrow.py
@@ -24,13 +24,11 @@ with the pyarrow library.
 """
 from __future__ import annotations
 
-import multiprocessing
 import os
 from abc import ABC, abstractmethod
+from concurrent.futures import Executor
 from functools import lru_cache, singledispatch
 from itertools import chain
-from multiprocessing.pool import ThreadPool
-from multiprocessing.sharedctypes import Synchronized
 from typing import (
     TYPE_CHECKING,
     Any,
@@ -38,6 +36,7 @@ from typing import (
     Dict,
     Generic,
     Iterable,
+    Iterator,
     List,
     Optional,
     Set,
@@ -124,6 +123,7 @@ from pyiceberg.types import (
     TimeType,
     UUIDType,
 )
+from pyiceberg.utils.concurrent import ManagedThreadPoolExecutor, Synchronized
 from pyiceberg.utils.singleton import Singleton
 
 if TYPE_CHECKING:
@@ -799,7 +799,7 @@ def _task_to_table(
                 arrow_table = fragment_scanner.to_table()
 
         if limit:
-            with rows_counter.get_lock():
+            with rows_counter:
                 if rows_counter.value >= limit:
                     return None
                 rows_counter.value += len(arrow_table)
@@ -811,12 +811,12 @@ def _task_to_table(
             return None
 
 
-def _read_all_delete_files(fs: FileSystem, pool: ThreadPool, tasks: Iterable[FileScanTask]) -> Dict[str, List[ChunkedArray]]:
+def _read_all_delete_files(fs: FileSystem, executor: Executor, tasks: Iterable[FileScanTask]) -> Dict[str, List[ChunkedArray]]:
     deletes_per_file: Dict[str, List[ChunkedArray]] = {}
     unique_deletes = set(chain.from_iterable([task.delete_files for task in tasks]))
     if len(unique_deletes) > 0:
-        deletes_per_files: List[Dict[str, ChunkedArray]] = pool.starmap(
-            func=_read_deletes, iterable=[(fs, delete) for delete in unique_deletes]
+        deletes_per_files: Iterator[Dict[str, ChunkedArray]] = executor.map(
+            lambda args: _read_deletes(*args), [(fs, delete) for delete in unique_deletes]
         )
         for delete in deletes_per_files:
             for file, arr in delete.items():
@@ -870,15 +870,14 @@ def project_table(
         id for id in projected_schema.field_ids if not isinstance(projected_schema.find_type(id), (MapType, ListType))
     }.union(extract_field_ids(bound_row_filter))
 
-    rows_counter = multiprocessing.Value("i", 0)
-
-    with ThreadPool() as pool:
-        deletes_per_file = _read_all_delete_files(fs, pool, tasks)
+    with ManagedThreadPoolExecutor() as executor:
+        rows_counter = executor.synchronized(0)
+        deletes_per_file = _read_all_delete_files(fs, executor, tasks)
         tables = [
             table
-            for table in pool.starmap(
-                func=_task_to_table,
-                iterable=[
+            for table in executor.map(
+                lambda args: _task_to_table(*args),
+                [
                     (
                         fs,
                         task,

--- a/python/pyiceberg/table/__init__.py
+++ b/python/pyiceberg/table/__init__.py
@@ -17,11 +17,11 @@
 from __future__ import annotations
 
 from abc import ABC, abstractmethod
+from concurrent.futures import ThreadPoolExecutor
 from dataclasses import dataclass
 from enum import Enum
 from functools import cached_property
 from itertools import chain
-from multiprocessing.pool import ThreadPool
 from typing import (
     TYPE_CHECKING,
     Any,
@@ -773,11 +773,11 @@ class DataScan(TableScan):
         data_entries: List[ManifestEntry] = []
         positional_delete_entries = SortedList(key=lambda entry: entry.data_sequence_number or INITIAL_SEQUENCE_NUMBER)
 
-        with ThreadPool() as pool:
+        with ThreadPoolExecutor() as executor:
             for manifest_entry in chain(
-                *pool.starmap(
-                    func=_open_manifest,
-                    iterable=[
+                *executor.map(
+                    lambda args: _open_manifest(*args),
+                    [
                         (
                             io,
                             manifest,

--- a/python/pyiceberg/utils/concurrent.py
+++ b/python/pyiceberg/utils/concurrent.py
@@ -1,0 +1,67 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=redefined-outer-name,arguments-renamed,fixme
+"""Concurrency concepts that support multi-threading."""
+import threading
+from concurrent.futures import Executor, ThreadPoolExecutor
+from contextlib import AbstractContextManager
+from typing import Any, Generic, TypeVar
+
+from typing_extensions import Self
+
+T = TypeVar("T")
+
+
+class Synchronized(Generic[T], AbstractContextManager):  # type: ignore
+    """A context manager that provides concurrency-safe access to a value."""
+
+    value: T
+    lock: threading.Lock
+
+    def __init__(self, value: T, lock: threading.Lock):
+        super().__init__()
+        self.value = value
+        self.lock = lock
+
+    def __enter__(self) -> T:
+        """Acquires a lock, allowing access to the wrapped value."""
+        self.lock.acquire()
+        return self.value
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        """Releases the lock, allowing other threads to access the value."""
+        self.lock.release()
+
+
+class ManagedExecutor(Executor):
+    """An executor that provides synchronization."""
+
+    def synchronized(self, value: T) -> Synchronized[T]:
+        raise NotImplementedError
+
+
+class ManagedThreadPoolExecutor(ThreadPoolExecutor, ManagedExecutor):
+    """A thread pool executor that provides synchronization."""
+
+    def __enter__(self) -> Self:
+        """Returns the executor itself as a context manager."""
+        super().__enter__()
+        return self
+
+    def synchronized(self, value: T) -> Synchronized[T]:
+        lock = threading.Lock()
+        return Synchronized(value, lock)

